### PR TITLE
Adding tests for build process

### PR DESCRIPTION
**Pull request by lukasoppermann**
## Summary
This PR adds tests for all or most functions within the `config` folder. All those functions are used when building color design tokens and are essential for the build so we should make sure they work as expected.

## Notable changes
- [x] adds tests for parser
- [x] adds tests for transformer
- [x] adds tests for filter
- [x] adds tests for utilities
- [x] adds tests for formats

### CODE REVIEW

File: __tests__/config/formats/typescript-export-definition.test.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
+import {typescriptExportDefinition} from '../../../config/formats/typescript-export-defition'
```
Comment (rezrah) [inline __tests__/config/formats/typescript-export-definition.test.ts:1]: Heads up, typo in the original filename. 

```suggestion
import {typescriptExportDefinition} from '../../../config/formats/typescript-export-definition'
```

File: __tests__/config/formats/typescript-export-definition.test.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        yellow: {
+          value: '#FF0000'
+        }
+      },
+      size: {
+        default: {
+          value: 16,
+          $type: 'dimension'
+        },
+        // px: {
```
Comment (rezrah) [inline __tests__/config/formats/typescript-export-definition.test.ts:20]: Are you planning to remove these?

File: __tests__/buildJs/pathToDotNotation.test.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-const pathToDotNotation = require('../build').pathToDotNotation
+const pathToDotNotation = require('../../build').pathToDotNotation
```
Comment (rezrah) [inline __tests__/buildJs/pathToDotNotation.test.js:1]: general feedback / question here, but could we start co-locating these tests to the source files? We do this in our other Primer repo's, and think it should apply here too.

File: __tests__/config/formats/typescript-export-definition.test.ts (hunk 2 above)
Comment (lukasoppermann) [reply __tests__/config/formats/typescript-export-definition.test.ts:20]: No I am planning on re-enabling them once we move the size stuff over to this build process.

File: __tests__/buildJs/pathToDotNotation.test.js (hunk 3 above)
Comment (lukasoppermann) [reply __tests__/buildJs/pathToDotNotation.test.js:1]: Those three tests will be removed once we move everything from `build.js` into the new build file.

So if we talk about colocation it would be for the config folder. I am not 100% sure I love the idea, as this either means we have double the files e.g. `16` instead of `8` which makes folders a bit hard to parse or, we need to put every transformer, etc. into a subfolder just to co-locate the test, which also makes it hard to work.

With components it feels more natural as you have the css module there as well.

Maybe we can discuss this in a call as this may be a bit of a longer conversation. 😉

Comment (rezrah) [reply __tests__/buildJs/pathToDotNotation.test.js:1]: yeah, I can see why you'd want to keep these build.js specific tests here for now 👍 . I think the others _could_ move though. E.g. `__tests__/config/formats/typescript-export-definition.test.ts` can be colocated to `config/formats/typescript-export-definition.ts` and there's _less_ folder creation and traversal and those tests still pass fine. Might be missing something here, so happy to sync on it.

File: __tests__/buildJs/pathToKebabCase.test.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-const pathToKebabCase = require('../build').pathToKebabCase
+const pathToKebabCase = require('../../build').pathToKebabCase
```
Comment (rezrah) [inline __tests__/buildJs/pathToKebabCase.test.js:1]: While we're doing this.. can we switch to restructuring rather than assignment through notation?

```suggestion
const {pathToKebabCase} = require('../../build')
```

File: __tests__/config/filters/isSource.test.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+      value: 'blue is not source',
+      isSource: false
+    },
+    {
+      value: 'yellow is not source'
+    }
+  ] as StyleDictionary.TransformedToken[]
+  it('filters out non-source items', () => {
+    expect(items.filter(isSource)).toStrictEqual([items[0]])
+  })
```
Comment (rezrah) [inline __tests__/config/filters/isSource.test.ts:20]: A suggestion to improve tests like this.. 

You're asserting against a very specific position in the test fixture, but that value can change at any time and invalidate the test. It also doesn't capture unhappy paths or edge cases where multiple  truthy results are present in the fixture. 

```suggestion
  const result = items.filter(isSource)
  const expectedValue = items.filter(item => item.isSource)
  
  it('filters out non-source items', () => {
    expect(result).toStrictEqual(expectedValue)
  })
```

This should still achieve the same thing but now isn't dependent on the fixture data being correct, but also keeps `isSource` honest against regressions.

Ideally this would be done with property testing paradigms, but this should work in the interim.

File: __tests__/config/formats/javascript-commonJs.test.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  it('Formats tokens without prefix', () => {
+    const output = `exports.default = {
+  colors: {
+    red: "#FF0000",
+  },
+};
+`
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore: fake values to test formatter
```
Comment (rezrah) [inline __tests__/config/formats/javascript-commonJs.test.ts:48]: Worth stubbing these out? Could make them available globally as we seem to be using these in a few different tests.

File: __tests__/config/parsers/w3c-json-parser.test.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      color: {
+        comment: 'a red color',
+        value: 'red',
+        $type: 'color',
+        alpha: 0.5
+      }
+    })
+  })
+
+  it('parses empty json', () => {
```
Comment (rezrah) [inline __tests__/config/parsers/w3c-json-parser.test.ts:28]:
```suggestion
  it('throws an error while parsing JSON without required properties', () => {
```

File: __tests__/config/parsers/w3c-json-parser.test.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+      })
+    ).toStrictEqual({
+      color: {
+        comment: 'a red color',
+        value: 'red',
+        $type: 'color',
+        alpha: 0.5
+      }
+    })
+  })
```
Comment (rezrah) [inline __tests__/config/parsers/w3c-json-parser.test.ts:26]: nit for readability

```suggestion
 it('parses valid json5', () => {
    const mockJSON = {
      contents: `{
        "color": {
          // comment
          "$value": 'red',
          "$description": "a red color",
          "$type": "color",
          alpha: 0.5
        }
      }`,
      filePath: ''
    }
    
    const result = w3cJsonParser.parse(mockJSON)

    const expectedOutput = {
      color: {
        comment: 'a red color',
        value: 'red',
        $type: 'color',
        alpha: 0.5
      }
    }

    expect(result).toStrictEqual(expectedOutput)
  })
```

File: __tests__/config/formats/javascript-commonJs.test.ts (hunk 6 above)
Comment (lukasoppermann) [reply __tests__/config/formats/javascript-commonJs.test.ts]: Yes, we could. Either we create a dataset somewhere or we create a function that adds all the other stuff to the tokens so you only define what is needed for the test. I'll look into it.

File: config/formats/scss-mixin-css-variables.test.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      & {
+        --red: transformedValue;
+      }
+    }`,
+      {parser: 'scss', printWidth: 500}
+    )
+    expect(scssMixinCssVariables(input)).toStrictEqual(expectedOutput)
+  })
+
+  it('Formats throws error if options.mixinName is missing', () => {
```
Comment (rezrah) [inline config/formats/scss-mixin-css-variables.test.ts:25]: should this be:

```suggestion
  it('throws and error if options.mixinName is missing
```

File: config/formats/typescript-export-definition.test.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
+import {typescriptExportDefinition} from './typescript-export-definition'
+import {getMockFormatterArguments} from '../../test-utilities/getMockFormatterArguments'
+import {format} from 'prettier'
+import {getMockDictionary} from '../../test-utilities/getMockDictionary'
+import {getMockToken} from '../../test-utilities/getMockToken'
+
+describe('Format: Tyepscript definitions', () => {
```
Comment (rezrah) [inline config/formats/typescript-export-definition.test.ts:7]:
```suggestion
describe('Format: TypeScript definitions', () => {
```

File: config/formats/typescript-export-definition.ts [hunk 11]
```diff
@@ -36,22 +43,24 @@ const getTokenType = (tokenTypesPath: string): string => {
...
     case 'color':
       if (value[0] === '#') {
         return 'ColorHex'
       }
       return 'string'
     case 'dimension':
-      if (value.substring(value.length - 3) === 'rem') {
-        return 'SizeRem'
-      }
+      if (typeof value === 'number') return 'number'
```
Comment (rezrah) [inline config/formats/typescript-export-definition.ts:54]: value seems to be typed as a string.. should this be !isNan instead? to account for values like "2"? or should the type defs be updates to accept string and number.

File: config/utilities/jsonToNestedValue.test.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        noValue: {
+          string: 'string',
+          number: 12
+        }
+      }
+    }
+  }
+
+  it('it returns only value property as value of object', () => {
+    expect(jsonToNestedValue(obj['parent']['subparent']['child1'])).toStrictEqual('#111')
```
Comment (rezrah) [inline config/utilities/jsonToNestedValue.test.ts:28]: just curious.. do you need to use index here? makes it a little harder to read... this should work same...

```suggestion
   expect(jsonToNestedValue(obj.parent.subparent.child1)).toBe('#111')
```

File: config/utilities/upperCaseFirstCharacter.test.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+describe('Utilities: upperCaseFirstCharacter', () => {
+  it('it transforms all lowercase word', () => {
+    expect(upperCaseFirstCharacter('primer')).toStrictEqual('Primer')
+  })
+
+  it('it transforms all lowercase sentence (words with spaces)', () => {
+    expect(upperCaseFirstCharacter('primer design token')).toStrictEqual('Primer design token')
+  })
+
+  it('it transforms all uppercase word', () => {
```
Comment (rezrah) [inline config/utilities/upperCaseFirstCharacter.test.ts:12]: is this a more accurate test name?
```suggestion
it('it preserves casing for words that are already all uppercased', () => {
```

File: tsconfig.json [hunk 14]
```diff
@@ -13,6 +13,6 @@
     "declaration": true,
     "resolveJsonModule": true
   },
-  "include": ["src/", "script/", "docs/", "config/types/"],
+  "include": ["src/", "script/", "docs/", "config/types/", "config/**/*.test.ts", "test-utilities/"],
```
Comment (rezrah) [inline tsconfig.json:16]: could test-utilities go under src instead? This `includes` array is getting quite large and might make it less predictable for others when it comes to debugging.

File: config/formats/typescript-export-definition.ts (hunk 11 above)
Comment (lukasoppermann) [reply config/formats/typescript-export-definition.ts:54]: I think updating the type definition would be correct here. But technically a `[dimension` must be a string with `px` or `rem`](https://design-tokens.github.io/community-group/format/#dimension) so maybe we can just scrap this line. What do you think?

Comment (rezrah) [reply config/formats/typescript-export-definition.ts:54]: yeah scrap it if it never gets evaluated 👍

**Review by rezrah — APPROVED** [review]
Lovely work as always @lukasoppermann ✨ 

Left some minor feedback for you, but otherwise good to go 🚀

File: tsconfig.json (hunk 14 above)
Comment (lukasoppermann) [reply tsconfig.json:16]: Yes, does that make sense? I think we wanted to [move tokens into source (`src/`)](https://github.com/github/primer/pull/1309/files?short_path=722b130#diff-722b130da8c969a084abdcb34977e310c071d1af7f0458ef6f8a35a7155c1a4c) so would this still make sense then? And if yes would we move `script/` into source as well? And `config`?

Is there a common way to structure this at github?

Comment (rezrah) [reply tsconfig.json:16]: not an enforced convention AFAIK, but you typically not want to overload the includes array and rather use it as a catch-all. Also you have config/types in there. Not sure what that is exactly, but just on naming it may be worth putting that under the [typeRoots](https://www.typescriptlang.org/tsconfig#typeRoots) property instead?
